[Multiplatform] Add support for multi platform conditions for source and resource files

alexanderwe opened the pull request:
This PR tries to add the platform filter approach to source and resource files. 

Addresses https://github.com/tuist/tuist/issues/5515

Short description 📝

The recently released multi platform support in tuist currently only supports setting platform filters on the target and target dependency level. This PR tries to enhance the current implementation by also making the filters available for source and resource files. 


Todos: 

- [x] Extract `TargetDependency.Condition` to a more generic place and probably also rename it 
- [x] Adapt tests
- [x] Add code documentation
- [x] Remove changes only applicable for debugging 
- [x] Linting

How to test the changes locally 🧐

All tests should pass

### Contributor checklist ✅

- [x] The code has been linted using run `make workspace/lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [ ] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [x] The code architecture and patterns are consistent with the rest of the codebase
- [x] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

waltflanagan commented on Sources/TuistGenerator/Generator/BuildPhaseGenerator.swift:
```diff
@@ -217,6 +218,7 @@ final class BuildPhaseGenerator: BuildPhaseGenerating {
     func generateSourcesBuildPhase(
         files: [SourceFile],
         coreDataModels: [CoreDataModel],
+        target: Target,
```
We may not need to address this with this work, but seeing this needing to be passed in is making me think that we may be able to move the `applicableTo` logic into the mapping phase now.    For this PR i think this is fine but something we can think on for future work.

waltflanagan commented on Sources/TuistGenerator/Generator/WorkspaceDescriptorGenerator.swift:
```diff
@@ -37,7 +37,7 @@ final class WorkspaceDescriptorGenerator: WorkspaceDescriptorGenerating {
         var projectGenerationContext: ExecutionContext
         static var `default`: Config {
-            Config(projectGenerationContext: .concurrent)
+            Config(projectGenerationContext: .serial)
```
Reminder: this is for debugging.

alexanderwe replied:
Thanks a lot for the reminder. Remove this code now

waltflanagan commented on Tests/TuistGeneratorTests/Generator/BuildPhaseGeneratorTests.swift:
```diff
@@ -632,6 +703,68 @@ final class BuildPhaseGeneratorTests: TuistUnitTestCase {
+        var buildFiles = try XCTUnwrap(resourceBuildPhase.files)
+
+        // sorting to get some determinism in, sort alphabetically, ascending, (uppercase before lowercase)
+        try buildFiles.sort { prev, suc in
```
would it be more straight forward to set up `expected` as a dictionary of filenames -> expected platform filters and just do a `for` loop over the `buildFiles` and look up that the filters match?  This would avoid having order matter and if we're doing an individual `XCTAssetEqual` within the `for` loop it would be more granular signal if something failed.

alexanderwe replied:
Thanks a lot for the suggestion - I added the approach you mentioned now

waltflanagan commented:
regarding naming, there are quite a few spots this shows up:
Dependencies
Copy bundle resources
Source Files
Headers
Run build tool plugin
Plist values for Minimum deployment versions, display name, build number, app icon

For now i think `InclusionCondition` makes sense.

We could also flip things a little and call it a `PlatformCondition` and certain types use `PlatformCondition` which only specifies platform filters.  This would be in contrast to the type that would eventually be used for those plist values mentioned above since those can be configured conditionally with respect to both platform and build configuration (debug vs release).

The thought is that various types (`TargetDependency`, `SourceFileGlob`, etc)  would have a condition that would depend based on the contents of the condition.  Right now our conditions are *just* the platform, so naming there would be `PlatformCondition`.   The thought is that the name describes the contents of the Condition, not the usage.

alexanderwe commented:
Thanks a lot for the input. I think then it depends on what we want to do with the type. As far as I have understood it the type could also be used for other conditions not only platform related ones. 

Maybe a generic name makes more sense then. Or do you think if we need another condition in the future a new type should be created, then I would go for`PlatformCondition`. If we would like to capture the concept of the condition in a single type and extend it later I would stick to the `InclusionCondition` way of naming.

waltflanagan commented:
@alexanderwe yea my thought was that we would create a new kind in the future if there is a condition that encompasses more than just platform and that type would be used where needed.  I have a small preference for `PlatformCondition`.  We could also use the callsite where its used to clarify what behavior the condition changes.  perhaps on dependencies it could be `linkingCondition: PlatformCondition`,  sources: `compilationCondition: PlatformCondition`, and resources `inclusionCondition: PlatformCondition`

alexanderwe commented:
@waltflanagan I like it. I moved `TargetDependency.Condition` to `PlatformCondition` now. I changed the parameter names for source and resource files. I kept the parameter name for `TargetDependency` though, since I was a little bit hesitant to change it because of probably introducing a breaking change of new manifest files that would use the  parameter name `condition`. But if it's okay to introduce such a small breaking change, I can of course change this as well.

waltflanagan commented on Sources/TuistDependencies/SwiftPackageManager/Utils/PackageInfoMapper.swift:
```diff
@@ -1437,10 +1437,10 @@ extension PackageInfoMapper {
     struct OnlyConditionsWithUnsupportedPlatforms: Error {}
 
-    /// Map from a package condition to ProjectDescription.TargetDependency.Condition
+    /// Map from a package condition to ProjectDescription.InclusionCondition
```
This comment should be updated.
```suggestion
    /// Map from a package condition to ProjectDescription.PlatformCondition
```

waltflanagan commented on Sources/ProjectDescription/PlatformCondition.swift:
```diff
@@ -0,0 +1,18 @@
+/// A condition applied to an "entity" allowing it to only be used in certain circumstances
+public struct PlatformCondition: Codable, Hashable, Equatable {
+    public let platformFilters: Set<PlatformFilter>
+    /// For internal use only. use `.when` to ensure we can not have a `InlcusionCondition` with an empty set of filters.
```
```suggestion
    /// For internal use only. use `.when` to ensure we can not have a `PlatformCondition` with an empty set of filters.
```

danyf90 commented on Sources/ProjectDescription/SourceFilesList.swift:
```diff
@@ -14,36 +14,54 @@ public struct SourceFileGlob: Codable, Equatable {
     public let codeGen: FileCodeGen?
 
+    /// Source file condition for compilation
+    public let compilationCondition: PlatformCondition?
```
Why not
```suggestion
    public let platformCondition: PlatformCondition?
```
?

alexanderwe commented:
@waltflanagan @danyf90 Thanks a lot for helping and merging this feature in ! 👍
